### PR TITLE
Fixed long-standing bug in balanced_tree.jl that corrupted SortedMult…

### PR DESCRIPTION
**Pull request by StephenVavasis**
…iDict

for certain insertions.

### CODE REVIEW

**Comment by oxinabox** [comment]
It will be a little while before I can review.

Are there still tests to be added?

**Comment by StephenVavasis** [comment]
I can't think of any other useful tests.  It's kind of surprising that the existing test set did not catch this error.  If it makes sense, I can add the OP's example to the tests.  Let me know if I can help with the code review.

**Comment by oxinabox** [comment]
> If it makes sense, I can add the OP's example to the tests.

I think it does, just as a regression test to be sure this exact error case doesn't happen again

**Comment by StephenVavasis** [comment]
I added the new test, but I am unable to load my branch to test the new test because apparently I am using Pkg incorrectly.  See the error message below. I got the same error from Pkg.develop.  Sorry, I still do not have a good grasp on Pkg.

```julia
julia> Pkg.add(url="https://github.com/StephenVavasis/DataStructures.jl#fix_sortedmultidict_bug")
     Cloning git-repo `https://github.com/StephenVavasis/DataStructures.jl#fix_sortedmultidict_bug`
    Updating git-repo `https://github.com/StephenVavasis/DataStructures.jl#fix_sortedmultidict_bug`
ERROR: could not find project file in package at `https://github.com/StephenVavasis/DataStructures.jl#fix_sortedmultidict_bug` maybe `subdir` needs to be specified
```

**Comment by oxinabox** [comment]
I would just use git.
git clone your branch then run it with `julia --project="."` then `] test`.

**Comment by StephenVavasis** [comment]
OK, that worked, thanks!  So the modified test set is good to go.

**Comment by StephenVavasis** [comment]
Any progress on this?  The bug is serious and would affect all users of SortedMultiDict.  Thanks.

**Comment by StephenVavasis** [comment]
By the way, here is the official way to use Pkg to add a branch from github:

``` 
Pkg.add(PackageSpec(url="https://github.com/StephenVavasis/DataStructures.jl", rev="fix_sortedmultidict_bug"))
```
I found out how to do this here: https://github.com/JuliaLang/Pkg.jl/issues/2852

File: src/balanced_tree.jl [hunk 1]
```diff
@@ -395,134 +396,122 @@ function Base.insert!(t::BalancedTree23{K,D,Ord}, k, d, allowdups::Bool) where {
...
-                                          oldtreenode.splitkey1, oldtreenode.splitkey1)
-            righttreenodenew = TreeNode{K}(newchild, oldtreenode.child3, 0,
-                                           oldtreenode.parent,
-                                           oldtreenode.splitkey2, oldtreenode.splitkey2)
-            whichp = 2
+    ## change the 2-node to a 3-node with the new child.
+
+    while true
+
+
```
Comment (oxinabox) [inline src/balanced_tree.jl:426]:
```suggestion
```

File: src/balanced_tree.jl [hunk 2]
```diff
@@ -395,134 +396,122 @@ function Base.insert!(t::BalancedTree23{K,D,Ord}, k, d, allowdups::Bool) where {
...
-                                           oldtreenode.parent,
-                                           minkeynewchild, minkeynewchild)
-            minkeynewchild = oldtreenode.splitkey2
-            whichp = 2
+            throw(AssertionError("Tree structure is corrupted 1"))
+        end
+
+        # Two cases: either we need to split the tree node
+        # if newchild4>0 else we convert a 2-node to a 3-node
+        # if newchild4==0
```
Comment (oxinabox) [inline src/balanced_tree.jl:464]:
```suggestion
        # Two cases: 
        # if newchild4>0  we need to split the tree node
        # elseif newchild4==0  we convert a 2-node to a 3-node
```

File: src/balanced_tree.jl [hunk 3]
```diff
@@ -647,6 +637,7 @@ function compareInd(t::BalancedTree23, i1::Int, i2::Int)
         p1 = t.tree[i1a].parent
         p2 = t.tree[i2a].parent
         @invariant_support_statement curdepth -= 1
+        curdepth -= 1
```
Comment (oxinabox) [inline src/balanced_tree.jl:640]:
```suggestion
```

File: src/balanced_tree.jl [hunk 4]
```diff
@@ -624,6 +613,7 @@ function compareInd(t::BalancedTree23, i1::Int, i2::Int)
     p1 = t.data[i1].parent
     p2 = t.data[i2].parent
     @invariant_support_statement curdepth = t.depth
+    curdepth = t.depth
```
Comment (oxinabox) [inline src/balanced_tree.jl:616]:
```suggestion
```

**Review by oxinabox — APPROVED** [review]
merge when happy

**Comment by StephenVavasis** [comment]
I don't think I have permission to merge.

**Comment by oxinabox** [comment]
Huh, I was sure you were a member.
You should be.
[Here is your invite](https://github.com/JuliaCollections)
Read and follow https://github.com/SciML/ColPrac

**Comment by StephenVavasis** [comment]
Thanks!  I did a squash-and-merge.  Can we also make a new release, since this is a major bug-fix?

**Comment by oxinabox** [comment]
Yes, I will backport it and release

**Pull request merged**